### DISCUSSION TITLE
Swell Has Been Down for nearly 10 hours

### DISCUSSION THREAD

**Discussion by osseonews** [swellstores/community#199] [Help] [answered]
Swell has been down for nearly 10 hours with no ability for anyone to use the cart or checkout functionality on our website. The API just keeps throwing errors. We can't even access https://swell.store/login. All of our API requests have been blocked by CORS? Why are our API requests which always worked getting blocked suddenly?

**Comment by osseonews** [comment]
We haven't changed a thing on our website in months. Just a basic Fetch to the cart via SwellJS is now blocked out of nowhere. What is going on?

`Access to fetch at 'https://OURSTORE.swell.store/api/cart, blocked by CORS policy: Response to preflight request doesn't pass access control check: No 'Access-Control-Allow-Origin' header is present on the requested resource.`

??? How are API requests blocked by CORS? The entire basis of using Swell is the ability to call API requests with SwellJs in the browser. How can this suddenly get blocked by CORS errors? This means the service is simply unusable. You need to turn off the CORS immediately on your server and simply allow all incoming requests, as was always the case when you request via SwellJS with the API + key. It is not acceptable for an API-driven platform to block incoming requests.

**Comment by osseonews** [comment]
BTW, the dashboard has also been down for hours. Supposedly a fix has been deployed, but this is false. The dashboard still returns a 503 Service Unavailable and/or the following error: "error": {
"code": "server_error",
"message": "Client is not defined"
}

**Comment by ericingram** [comment]
@osseonews this was due to a major change (a rare but necessary effort) that unfortunately broke things. It looks like we are mostly in the clear now but continuing to work on reported issues that could be related.

**Comment by osseonews** [comment]
Well the issue is that because this major change broke things, your API was not responsive for over 12 hours, causing major downstream effects and new bugs. For example, we host our Nextjs build on Swell at Netlify and our Netlify function usage exceeded our monthly threshold by over 100% in 12 hours due to Swell's unresponsive api. Will Swell provide a credit for the losses associated with this major change that broke things?

**Reply by ericingram** [reply]
The API was responsive during this period, but specific functions were not working as expected.

Our enterprise SLA offers a credit based on downtime, but it's not intended to compensate for external costs.

**Reply by BelphegorPrime** [reply]
@osseonews are you using nextjs commerce?
If yes I raised a ticket here. 
https://github.com/vercel/commerce/issues/961

We where hosting on Vercel but in general it should be the same with netlify.

**Reply by osseonews** [reply]
Thanks for bringing this up. Good to know, at least that we are not the only site that experienced this. Yes we use NextJs Commerce and we raised the issue with Netlify. Haven't heard back yet. Received a massive bill due to the DDOS from Swell API.  Did Vercel bill for this also? BTW, I don't think Vercel really actively maintains NextJS Commerce anymore,  plus what transpired here was not an issue with Nextjs commerce at all. How could you have stopped this other than just taking down the entire website?

**Reply by BelphegorPrime** [reply]
I saw another guy being ddosed by a webhook, but if i understand it right, nextjs commerce is just polling the api.
So nextjs commerce should implement error handling accordingly.
For example in case of network errors doing a exponential backoff.

**Reply by osseonews** [reply]
What transpired had nothing to do with Nextjs commerce and nothing they have in the code is remotely responsible or could have prevented this.  NextjsCommerce is based on NextJs libraries. It uses useSwr for client-side interactions and also the basic Nextjs functionality of Incremental Static Generation (ISR) for pages. ISR  will attempt to revalidate or "poll" your pages at the interval you set. These are basic features of NextJs libraries, in general, and nothing to with NextJscommerce specifically. Neither useSwr or ISR will poll your website 8 million times in a few hours, as I'll explain below.  They cannot do the equivalent of a DDOS attack on your website with a failed API. It is just impossible for regular users. 

In general, as for specific intervals, useSwr will only "poll" your API when a user is interacting with their browser, and in that case,  useSwr will out of the box do polling roughly every minute, but you are free to change this interval to whatever you want or do no polling at all. However, if there is no user interacting with your website, it does not poll anything. As for ISR, if you look at nextjscommerce in StaticProps it is set to revalidate every hour (revalidate: 60 * 60, // Every hour).  Again, you can set ISR to whatever you want, but at worst Nextjs will revalidate every minute and if you didn't change the settings in NextjsCommerce it was for 1 hour per page.

So based on the above, with ISR set in Nextjs Commerce if an API is failing, you will at most have 24 requests over the span of a day per page (once per hour per page). So you would need to have over  1 million pages on your website to reach anywhere near the level of requests via ISR via a failed API with ISR that was reported on that day. I suspect you don't have 1 million pages. We certainly don't. Actually, we have around 200 pages, so a failed API would result in around 6,000 API requests for the entire day!  As for useSwr at most you would have around 2,000 requests per day per user. You would need 4,000 customers per minute using your website to achieve 8 million of requests on a broken API over a day. I suspect you don't have 4,000 customers a minute on your website or we wouldn't be chatting here :0). 

Now getting back to what happened here, if you look at your logs, none of the above is relevant. It wasn't user browsers that was causing anything, as you will see in the logs that it wasn't the useSwr driven pages that were getting hit. **The massive increase in API requests was occurring on [....page] route which is an ISR and should revalidate every hour or at worst every minute (again depends on your setting). However, in this case, the pages were getting hit every millisecond!** 

So overall, this was not a Nextjs issue at all. This was a DDOS style attack due to a rogue service. It was not the polling that was causing the DDOS, it was the "attacker", in this case Swell's malfunctioning API, that was firing requests at the website every millisecond for whatever reason. I don't have access to swell's servers to see why that happened, but I can speculate. Anyway, if Swell had anyone monitoring the service they would have seen this problem immediately and shut down the API. But, they didn't have anyone monitoring anything and/or didn't have anyone who understood what was going on monitoring it. I mean it is a trivial issue to fix actually if have a rogue API: You shut down your API servers until your figure out the bug! Was this done? No. In fact, it was worse, because despite us reporting this issue continuously during that time, they actually denied (and still deny) the reality of the situation and attempted to shift the blame to users. Classic "blame the victim" tactic when you don't want to take any responsibility.

**Comment by osseonews** [comment]
<img width="1087" alt="swell" src="https://user-images.githubusercontent.com/26583659/230487208-4ff916d3-3354-477f-bc10-a496462b1af2.png">

**Reply by osseonews** [reply]
So if there is an API outage, the API is responsive? How does that work exactly?

**Reply by ericingram** [reply]
It wasn't completely out. But there were specific functions that didn't behave normally. The verbiage on the status page was inaccurate I suppose.

**Reply by osseonews** [reply]
The verbiage on the status page was exactly correct. The API was out and wasn't working for nearly 12 hours. Your irresponsible upgrade broke things and caused massive damage that is still ongoing for us.

**Reply by oskajory** [reply]
> The verbiage on the status page was exactly correct. The API was out and wasn't working for nearly 12 hours. Your irresponsible upgrade broke things and caused massive damage that is still ongoing for us.

Hi osseonews, we had a team of 5 people working on building stores with Swell with an Enterprise agreement and can confidently say that the team was completely unresponsive when needed, they were untrustworthy, they advertised stuff they DID NOT HAVE that we relied on, they would not take any responsibility on any problems that they were at fault for with their platform. Our setup process was a nightmare from the start with them, and kept trying to push back saying our developers are incapable essentially when their documentation wasn't even complete or had half finished sections or missing sections.. We had to pull out due to the lack of stability, support, constant issues with the platform and the frustrations with how poor the honesty and reliability of their internal teams were. Not to mention their entire platform is built on NodeJS. Why is an entire Ecommerce platform built on NodeJS? No company should be pushing updates that can break existing sites - losing customers revenue. If that is the case it deems your platform unfit for commercial use. We even called them out for not following through with "enterprise level support" that we signed and agreed on and told us to just exercise cancellation.

**Reply by osseonews** [reply]
And I thought it was just me :0)

BTW, in terms of technology, I don't agree with the NodeJs comment. I think Node is excellent for e-commerce. However what I would say is that, ignoring all the other non-tech problems, the flaw here is building the entire platform on Mongo. Mongo is great for certain things, but terrible for others. Undoubtedly, it is unsuitable for certain critical pieces of an ecommerce engine on a larger scale, and I suspect, though I cant be sure, that alot of Swell's headaches are driven by Mongo limitations as related to certain aspects of e-commerce.

**Comment by bafomet88** [comment]
10 hours major breakdown that cuts down stores revenue, and hardly any reaction from the company. They even didn't send any e-mail, not to mention compensation proposition. That doesn't look good for the future, swell has lost a lot in my eyes, I'm seriously worried about the trustworthiness of this company.

**Comment by ericingram** [comment, marked as answer]
We are working on a full review of the changes and resulting outage to publish in full transparency, while also prioritizing fixes currently.

The system wasn't fully down, but many functions were definitely broken during that time. (Particularly hosted storefronts, hosted checkout, much of the storefront API, as well as the admin dashboard).

We're very sorry for the trouble caused, and you can expect more details soon.

**Reply by logeshswell** [reply]
https://github.com/orgs/swellstores/discussions/215
